File: C++/14-Backtracking/Algorithms/3-HamiltonianCycle/HamiltonianCyclePro.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <numeric>   // For iota
#include <limits>    // For numeric_limits
#include <chrono>    // For timing
#include <random>    // For std::mt19937 and std::uniform_int_distribution

using namespace std;
using namespace std::chrono;

// Function to check if a given vertex 'v' can be added to the current path
bool isSafe(int v, vector<int>& path, vector<vector<int>>& graph, int pos) {
    // Check if the vertex 'v' is adjacent to the last vertex in the path.
    if (graph[path[pos - 1]][v] == 0)
        return false;

    // Check if the vertex 'v' is already present in the current path.
    for (int i = 0; i < pos; i++)
        if (path[i] == v)
            return false;

    return true;
}

// 1. Basic Backtracking Approach
//   - Recursively explores all possible paths.
//   - Returns true if a Hamiltonian Cycle exists, otherwise false.
bool hamiltonianCycleBasic(vector<vector<int>>& graph, vector<int>& path, int pos) {
    int V = graph.size();

    // Base case: If all vertices are included in the path
    if (pos == V) {
        // Check if there is an edge from the last vertex to the starting vertex.
        if (graph[path[pos - 1]][path[0]] == 1)
            return true;
        else
            return false;
    }

    // Try different vertices as the next vertex in the path.
    for (int v = 0; v < V; v++) {
        if (isSafe(v, path, graph, pos)) {
            path[pos] = v;
            // Recursively explore the remaining path.
            if (hamiltonianCycleBasic(graph, path, pos + 1))
                return true;
            // Backtrack: If adding vertex 'v' doesn't lead to a solution, remove it.
            path[pos] = -1;
        }
    }

    // If no vertex can be added to the path, return false.
    return false;
}

// Function to find a Hamiltonian Cycle using the basic backtracking approach.
bool findHamiltonianCycleBasic(vector<vector<int>>& graph) {
    int V = graph.size();
    vector<int> path(V, -1); // Initialize path with -1 (no vertex assigned)
    path[0] = 0;             // Start the path from vertex 0

    // Start the recursion from the second vertex (index 1 in path).
    if (hamiltonianCycleBasic(graph, path, 1) == false) {
        cout << "Hamiltonian Cycle does not exist (Basic)." << endl;
        return false;
    } else {
        cout << "Hamiltonian Cycle found (Basic): ";
        for (int vertex : path)
            cout << vertex << " ";
        cout << path[0] << endl; // Print the starting vertex to complete the cycle.
        return true;
    }
}

// 2. Backtracking with Forward Checking
//    - Improves basic backtracking by checking future possibilities.
//    - If a vertex has no unvisited neighbors, the current path is invalid.
bool hamiltonianCycleForwardChecking(vector<vector<int>>& graph, vector<int>& path, int pos, vector<bool>& visited) {
    int V = graph.size();

    if (pos == V) {
        if (graph[path[pos - 1]][path[0]] == 1)
            return true;
        else
            return false;
    }

    for (int v = 0; v < V; v++) {
        if (isSafe(v, path, graph, pos) && !visited[v]) {
            // Forward checking:
            // Check if adding 'v' will leave any unvisited node with no path
            bool valid = true;
            visited[v] = true; // Mark v as visited for the check
            int unvisitedCount = 0;
            for (int u = 0; u < V; ++u) {
                if (!visited[u]) {
                    unvisitedCount++;
                    bool hasNeighbor = false;
                    for (int neighbor = 0; neighbor < V; ++neighbor) {
                        if (graph[u][neighbor] == 1 && !visited[neighbor]) {
                            hasNeighbor = true;
                            break;
                        }
                    }
                    if (!hasNeighbor) {
                        valid = false;
                        break;
                    }
                }
            }

            if (valid) {
                path[pos] = v;
                if (hamiltonianCycleForwardChecking(graph, path, pos + 1, visited))
                    return true;
                path[pos] = -1;
            }
            visited[v] = false; // backtrack
        }
    }
    return false;
}

bool findHamiltonianCycleForwardChecking(vector<vector<int>>& graph) {
    int V = graph.size();
    vector<int> path(V, -1);
    path[0] = 0;
    vector<bool> visited(V, false);
    visited[0] = true;

    if (hamiltonianCycleForwardChecking(graph, path, 1, visited) == false) {
        cout << "Hamiltonian Cycle does not exist (Forward Checking)." << endl;
        return false;
    } else {
        cout << "Hamiltonian Cycle found (Forward Checking): ";
        for (int vertex : path)
            cout << vertex << " ";
        cout << path[0] << endl;
        return true;
    }
}

// 3. Backtracking with Degree Ordering
//    - Prioritizes vertices with higher degrees.
//    - Improves performance in some graphs by exploring more promising paths first.
bool hamiltonianCycleDegreeOrdering(vector<vector<int>>& graph, vector<int>& path, int pos, vector<bool>& visited, vector<int>& degree) {
    int V = graph.size();

    if (pos == V) {
        if (graph[path[pos - 1]][path[0]] == 1)
            return true;
        else
            return false;
    }

    // Create a vector of unvisited vertices with their degrees.
    vector<pair<int, int>> unvisitedVertices; // (vertex, degree)
    for (int v = 0; v < V; ++v) {
        if (!visited[v] && isSafe(v, path, graph, pos)) { // Added isSafe check here
            unvisitedVertices.push_back({v, degree[v]});
        }
    }
    // Sort unvisited vertices in descending order of their degrees.
    sort(unvisitedVertices.begin(), unvisitedVertices.end(), [](const auto& a, const auto& b) {
        return a.second > b.second;
    });

    for (const auto& vertexDegreePair : unvisitedVertices) {
        int v = vertexDegreePair.first;
        if (isSafe(v, path, graph, pos)) { // Check if it is safe to visit the node.
            path[pos] = v;
            visited[v] = true;
            if (hamiltonianCycleDegreeOrdering(graph, path, pos + 1, visited, degree))
                return true;
            path[pos] = -1;
            visited[v] = false;
        }
    }
    return false;
}

bool findHamiltonianCycleDegreeOrdering(vector<vector<int>>& graph) {
    int V = graph.size();
    vector<int> path(V, -1);
    path[0] = 0;
    vector<bool> visited(V, false);
    visited[0] = true;
    vector<int> degree(V, 0);

    // Calculate the degree of each vertex.
    for (int i = 0; i < V; ++i) {
        for (int j = 0; j < V; ++j) {
            degree[i] += graph[i][j];
        }
    }

    if (hamiltonianCycleDegreeOrdering(graph, path, 1, visited, degree) == false) {
        cout << "Hamiltonian Cycle does not exist (Degree Ordering)." << endl;
        return false;
    } else {
        cout << "Hamiltonian Cycle found (Degree Ordering): ";
        for (int vertex : path)
            cout << vertex << " ";
        cout << path[0] << endl;
        return true;
    }
}

// 4. Backtracking with Warnsdorff's Rule (for Knight's Tour) - Adapted for Hamiltonian Cycle
//    - Originally for Knight's Tour, but adapted to prioritize vertices with fewer neighbors.
//    - May not guarantee a solution for all graphs, but often performs well in sparse graphs.
bool hamiltonianCycleWarnsdorff(vector<vector<int>>& graph, vector<int>& path, int pos, vector<bool>& visited, vector<int>& degree) {
    int V = graph.size();

    if (pos == V) {
        if (graph[path[pos - 1]][path[0]] == 1)
            return true;
        else
            return false;
    }

    vector<pair<int, int>> unvisitedVertices; // (vertex, num_neighbors)
    for (int v = 0; v < V; ++v) {
        if (!visited[v] && isSafe(v, path, graph, pos)) {
            int numNeighbors = 0;
            for (int neighbor = 0; neighbor < V; ++neighbor) {
                if (graph[v][neighbor] == 1 && !visited[neighbor]) {
                    numNeighbors++;
                }
            }
            unvisitedVertices.push_back({v, numNeighbors});
        }
    }

    // Sort unvisited vertices in ascending order of number of unvisited neighbors.
    sort(unvisitedVertices.begin(), unvisitedVertices.end(), [](const auto& a, const auto& b) {
        return a.second < b.second;
    });

    for (const auto& vertexNeighborPair : unvisitedVertices) {
        int v = vertexNeighborPair.first;
        if (isSafe(v, path, graph, pos)) {
            path[pos] = v;
            visited[v] = true;
            if (hamiltonianCycleWarnsdorff(graph, path, pos + 1, visited, degree))
                return true;
            path[pos] = -1;
            visited[v] = false;
        }
    }
    return false;
}

bool findHamiltonianCycleWarnsdorff(vector<vector<int>>& graph) {
    int V = graph.size();
    vector<int> path(V, -1);
    path[0] = 0;
    vector<bool> visited(V, false);
    visited[0] = true;
    vector<int> degree(V, 0);

    for (int i = 0; i < V; ++i) {
        for (int j = 0; j < V; ++j) {
            degree[i] += graph[i][j];
        }
    }

    if (hamiltonianCycleWarnsdorff(graph, path, 1, visited, degree) == false) {
        cout << "Hamiltonian Cycle does not exist (Warnsdorff's)." << endl;
        return false;
    } else {
        cout << "Hamiltonian Cycle found (Warnsdorff's): ";
        for (int vertex : path)
            cout << vertex << " ";
        cout << path[0] << endl;
        return true;
    }
}

// 5. Randomized Backtracking
//    - Introduces randomness in vertex selection to explore diverse paths.
//    - Can help escape local optima and find solutions in complex graphs.
bool hamiltonianCycleRandomized(vector<vector<int>>& graph, vector<int>& path, int pos, vector<bool>& visited) {
    int V = graph.size();

    if (pos == V) {
        if (graph[path[pos - 1]][path[0]] == 1)
            return true;
        else
            return false;
    }

    vector<int> possibleVertices;
    for (int v = 0; v < V; v++) {
        if (isSafe(v, path, graph, pos) && !visited[v]) {
            possibleVertices.push_back(v);
        }
    }

    // Use std::shuffle instead of random_shuffle (deprecated in C++14, removed in C++17)
    if (!possibleVertices.empty()) { // Make sure possibleVertices is not empty
        std::random_device rd;
        std::mt19937 gen(rd());
        std::uniform_int_distribution<> dist(0, possibleVertices.size() - 1);

        for (int i = possibleVertices.size() - 1; i > 0; --i) {
            int j = dist(gen);
            std::swap(possibleVertices[i], possibleVertices[j]);
        }
    }

    for (int v : possibleVertices) {
        path[pos] = v;
        visited[v] = true;
        if (hamiltonianCycleRandomized(graph, path, pos + 1, visited))
            return true;
        path[pos] = -1;
        visited[v] = false;
    }
    return false;
}

bool findHamiltonianCycleRandomized(vector<vector<int>>& graph) {
    int V = graph.size();
    vector<int> path(V, -1);
    path[0] = 0;
    vector<bool> visited(V, false);
    visited[0] = true;

    if (hamiltonianCycleRandomized(graph, path, 1, visited) == false) {
        cout << "Hamiltonian Cycle does not exist (Randomized)." << endl;
        return false;
    } else {
        cout << "Hamiltonian Cycle found (Randomized): ";
        for (int vertex : path)
            cout << vertex << " ";
        cout << path[0] << endl;
        return true;
    }
}

int main() {
    // Example graph represented as an adjacency matrix.
    //  0 1 2 3
    //0 0 1 0 1
    //1 1 0 1 1
    //2 0 1 0 1
    //3 1 1 1 0
    vector<vector<int>> graph = {
        {0, 1, 0, 1},
        {1, 0, 1, 1},
        {0, 1, 0, 1},
        {1, 1, 1, 0}
    };
    // Expected output: 0 1 3 2 0

    cout << "Graph:" << endl;
    for (const auto& row : graph) {
        for (int val : row) {
            cout << val << " ";
        }
        cout << endl;
    }
    cout << endl;

    // Test each approach and measure execution time.
    auto start = high_resolution_clock::now();
    findHamiltonianCycleBasic(graph);
    auto stop = high_resolution_clock::now();
    auto duration = duration_cast<microseconds>(stop - start);
    cout << "Time taken by Basic: " << duration.count() << " microseconds" << endl << endl;

    start = high_resolution_clock::now();
    findHamiltonianCycleForwardChecking(graph);
    stop = high_resolution_clock::now();
    duration = duration_cast<microseconds>(stop - start);
    cout << "Time taken by Forward Checking: " << duration.count() << " microseconds" << endl << endl;

    start = high_resolution_clock::now();
    findHamiltonianCycleDegreeOrdering(graph);
    stop = high_resolution_clock::now();
    duration = duration_cast<microseconds>(stop - start);
    cout << "Time taken by Degree Ordering: " << duration.count() << " microseconds" << endl << endl;

    start = high_resolution_clock::now();
    findHamiltonianCycleWarnsdorff(graph);
    stop = high_resolution_clock::now();
    duration = duration_cast<microseconds>(stop - start);
    cout << "Time taken by Warnsdorff's: " << duration.count() << " microseconds" << endl << endl;

    start = high_resolution_clock::now();
    findHamiltonianCycleRandomized(graph);
    stop = high_resolution_clock::now();
    duration = duration_cast<microseconds>(stop - start);
    cout << "Time taken by Randomized: " << duration.count() << " microseconds" << endl << endl;

    return 0;
}

File: Java/14-Backtracking/Algorithms/3-HamiltonianCycle/HamiltonianCyclePro.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Scanner;

public class HamiltonianCycle {

    private static final int V = 5; // Number of vertices in the graph.  Can be changed.

    // 1. Basic Backtracking Approach
    //   - Simple, recursive backtracking to find a Hamiltonian Cycle.
    //   - Good for understanding the core concept.
    //   - Real-world application:  Route planning for a robot in a small, simple environment.
    public static boolean hasHamiltonianCycleBasic(int[][] graph) {
        int[] path = new int[V];
        Arrays.fill(path, -1); // Initialize path array
        path[0] = 0; // Start from vertex 0
        return hasHamiltonianCycleBasicUtil(graph, path, 1);
    }

    private static boolean hasHamiltonianCycleBasicUtil(int[][] graph, int[] path, int pos) {
        if (pos == V) {
            // Check if there is an edge from the last vertex to the starting vertex (0)
            return graph[path[pos - 1]][0] == 1;
        }

        for (int v = 0; v < V; v++) {
            // Check if v is a valid next vertex
            if (isSafeBasic(v, graph, path, pos)) {
                path[pos] = v;
                if (hasHamiltonianCycleBasicUtil(graph, path, pos + 1)) {
                    return true;
                }
                // Backtrack: Remove v from path and try different vertex
                path[pos] = -1;
            }
        }
        return false;
    }

    private static boolean isSafeBasic(int v, int[][] graph, int[] path, int pos) {
        // Check if the vertex v is adjacent to the previous vertex in the path
        if (graph[path[pos - 1]][v] == 0) {
            return false;
        }

        // Check if the vertex v is already included in the path
        for (int i = 0; i < pos; i++) {
            if (path[i] == v) {
                return false;
            }
        }
        return true;
    }


    // 2. Backtracking with Forward Checking
    //   - Improves upon the basic approach by adding a forward checking step.
    //   - Checks if a vertex has unvisited neighbors before adding it to the path.
    //   - Real-world application:  More efficient route planning for delivery trucks,
    //     where it's important to ensure all locations can be visited.
    public static boolean hasHamiltonianCycleForwardChecking(int[][] graph) {
        int[] path = new int[V];
        Arrays.fill(path, -1);
        path[0] = 0;
        boolean[] visited = new boolean[V];
        visited[0] = true; // Mark the starting node as visited
        return hasHamiltonianCycleForwardCheckingUtil(graph, path, 1, visited);
    }

    private static boolean hasHamiltonianCycleForwardCheckingUtil(int[][] graph, int[] path, int pos, boolean[] visited) {
        if (pos == V) {
            return graph[path[pos - 1]][0] == 1;
        }

        for (int v = 0; v < V; v++) {
            if (isSafeForwardChecking(v, graph, visited, path, pos)) {
                path[pos] = v;
                visited[v] = true; // Mark vertex v as visited
                if (hasHamiltonianCycleForwardCheckingUtil(graph, path, pos + 1, visited)) {
                    return true;
                }
                path[pos] = -1;
                visited[v] = false; // Backtrack: unmark visited
            }
        }
        return false;
    }

    private static boolean isSafeForwardChecking(int v, int[][] graph, boolean[] visited, int[] path, int pos) {
        if (graph[path[pos - 1]][v] == 0) {
            return false;
        }
        if (visited[v]) {
            return false;
        }

        // Forward checking: Check if v has any unvisited neighbors.  If not, it's a dead end.
        boolean hasUnvisitedNeighbor = false;
        for (int i = 0; i < V; i++) {
            if (graph[v][i] == 1 && !visited[i]) {
                hasUnvisitedNeighbor = true;
                break;
            }
        }
        if (pos == V - 1 && graph[v][0] != 1){
            return false;
        }
        return hasUnvisitedNeighbor || pos == V - 1; // Allow it if it is the last node.
    }


    // 3. Backtracking with Degree Ordering
    //   - Optimizes the search by considering vertices in increasing order of their degree.
    //   - The degree of a vertex is the number of edges connected to it.
    //   - Real-world application:  Useful in network routing, where nodes with fewer connections
    //     might be more constrained and should be considered earlier.
    public static boolean hasHamiltonianCycleDegreeOrdering(int[][] graph) {
        int[] path = new int[V];
        Arrays.fill(path, -1);
        path[0] = 0;
        boolean[] visited = new boolean[V];
        visited[0] = true;
        int[] degree = new int[V];
        for (int i = 0; i < V; i++) {
            for (int j = 0; j < V; j++) {
                degree[i] += graph[i][j];
            }
        }
        return hasHamiltonianCycleDegreeOrderingUtil(graph, path, 1, visited, degree);
    }

    private static boolean hasHamiltonianCycleDegreeOrderingUtil(int[][] graph, int[] path, int pos, boolean[] visited, int[] degree) {
        if (pos == V) {
            return graph[path[pos - 1]][0] == 1;
        }

        // Create an array of vertices and sort it based on degree.
        Integer[] vertices = new Integer[V];
        for (int i = 0; i < V; i++) {
            vertices[i] = i;
        }
        Arrays.sort(vertices, (a, b) -> degree[a] - degree[b]); // Sort by degree

        for (int v : vertices) {
            if (isSafeForwardChecking(v, graph, visited, path, pos)) { // Use Forward Checking
                path[pos] = v;
                visited[v] = true;
                if (hasHamiltonianCycleDegreeOrderingUtil(graph, path, pos + 1, visited, degree)) {
                    return true;
                }
                path[pos] = -1;
                visited[v] = false;
            }
        }
        return false;
    }


    // 4. Backtracking with Warnsdorff's Rule (for Knight's Tour)
    //   - A heuristic specifically designed for the Knight's Tour problem (a special case
    //     of the Hamiltonian Cycle problem on a chessboard).
    //   - Prioritizes moving to the square with the fewest possible next moves.
    //   - Real-world application:  Path planning for specialized robotic arms or movement
    //     in constrained grid-like environments.
    public static boolean hasHamiltonianCycleWarnsdorff(int[][] graph, int startX, int startY) {
        int[][] path = new int[V][2]; // Store (x, y) coordinates of the path
        // Initialize path
        path[0][0] = startX;
        path[0][1] = startY;

        boolean[][] visited = new boolean[V][V];
        visited[startX][startY] = true;

        return hasHamiltonianCycleWarnsdorffUtil(graph, path, 1, visited);
    }

    private static boolean hasHamiltonianCycleWarnsdorffUtil(int[][] graph, int[][] path, int pos, boolean[][] visited) {
        if (pos == V) {
            int startX = path[0][0];
            int startY = path[0][1];
            int endX = path[pos-1][0];
            int endY = path[pos-1][1];
            return graph[endX][startX] == 1;
        }

        int currX = path[pos - 1][0];
        int currY = path[pos - 1][1];

        // Get valid moves (neighbors)
        List<int[]> moves = getValidMovesWarnsdorff(currX, currY, visited, graph);

        // Sort moves using Warnsdorff's Rule (sort by degree of the neighbor)
        moves.sort((a, b) -> countNeighbors(a[0], a[1], visited, graph) - countNeighbors(b[0], b[1], visited, graph));

        for (int[] move : moves) {
            int nextX = move[0];
            int nextY = move[1];

            path[pos][0] = nextX;
            path[pos][1] = nextY;
            visited[nextX][nextY] = true;

            if (hasHamiltonianCycleWarnsdorffUtil(graph, path, pos + 1, visited)) {
                return true;
            }

            path[pos][0] = -1; // Backtrack
            path[pos][1] = -1;
            visited[nextX][nextY] = false;
        }

        return false;
    }

    private static List<int[]> getValidMovesWarnsdorff(int x, int y, boolean[][] visited, int[][] graph) {
        List<int[]> moves = new ArrayList<>();
        int[] dx = {1, 2, 2, 1, -1, -2, -2, -1};
        int[] dy = {2, 1, -1, -2, -2, -1, 1, 2};

        for (int i = 0; i < 8; i++) {
            int newX = x + dx[i];
            int newY = y + dy[i];
            if (newX >= 0 && newX < V && newY >= 0 && newY < V && !visited[newX][newY] && graph[x][newX] == 1) {
                moves.add(new int[]{newX, newY});
            }
        }
        return moves;
    }

    private static int countNeighbors(int x, int y, boolean[][] visited, int[][] graph) {
        int count = 0;
        int[] dx = {1, 2, 2, 1, -1, -2, -2, -1};
        int[] dy = {2, 1, -1, -2, -2, -1, 1, 2};
        for (int i = 0; i < 8; i++) {
            int newX = x + dx[i];
            int newY = y + dy[i];
            if (newX >= 0 && newX < V && newY >= 0 && newY < V && !visited[newX][newY] && graph[x][newX] == 1) {
                count++;
            }
        }
        return count;
    }



    // 5. Randomized Backtracking
    //   - Introduces randomness into the vertex selection process.
    //   - Can help to escape local optima and explore a wider range of possible solutions.
    //   - Real-world application:  Useful for generating diverse solutions, such as in
    //     artistic design or game level generation.
    public static boolean hasHamiltonianCycleRandomized(int[][] graph) {
        int[] path = new int[V];
        Arrays.fill(path, -1);
        path[0] = 0;
        boolean[] visited = new boolean[V];
        visited[0] = true;
        return hasHamiltonianCycleRandomizedUtil(graph, path, 1, visited);
    }

    private static boolean hasHamiltonianCycleRandomizedUtil(int[][] graph, int[] path, int pos, boolean[] visited) {
        if (pos == V) {
            return graph[path[pos - 1]][0] == 1;
        }

        List<Integer> candidates = new ArrayList<>();
        for (int v = 0; v < V; v++) {
            if (isSafeForwardChecking(v, graph, visited, path, pos)) { // Use forward checking
                candidates.add(v);
            }
        }

        // Randomly shuffle the candidates
        java.util.Collections.shuffle(candidates);

        for (int v : candidates) {
            path[pos] = v;
            visited[v] = true;
            if (hasHamiltonianCycleRandomizedUtil(graph, path, pos + 1, visited)) {
                return true;
            }
            path[pos] = -1;
            visited[v] = false;
        }
        return false;
    }



    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);

        // Example graph (adjacency matrix representation)
        int[][] graph = {
                {0, 1, 0, 1, 0},
                {1, 0, 1, 1, 1},
                {0, 1, 0, 0, 1},
                {1, 1, 0, 0, 1},
                {0, 1, 1, 1, 0},
        };

        System.out.println("Hamiltonian Cycle Detection");
        System.out.println("-----------------------------");
        System.out.println("Graph (Adjacency Matrix):");
        for (int i = 0; i < V; i++) {
            System.out.println(Arrays.toString(graph[i]));
        }
        System.out.println();

        // Basic Backtracking
        System.out.println("1. Basic Backtracking:");
        if (hasHamiltonianCycleBasic(graph)) {
            System.out.println("Hamiltonian Cycle exists (Basic).");
        } else {
            System.out.println("Hamiltonian Cycle does not exist (Basic).");
        }
        System.out.println();

        // Backtracking with Forward Checking
        System.out.println("2. Backtracking with Forward Checking:");
        if (hasHamiltonianCycleForwardChecking(graph)) {
            System.out.println("Hamiltonian Cycle exists (Forward Checking).");
        } else {
            System.out.println("Hamiltonian Cycle does not exist (Forward Checking).");
        }
        System.out.println();

        // Backtracking with Degree Ordering
        System.out.println("3. Backtracking with Degree Ordering:");
        if (hasHamiltonianCycleDegreeOrdering(graph)) {
            System.out.println("Hamiltonian Cycle exists (Degree Ordering).");
        } else {
            System.out.println("Hamiltonian Cycle does not exist (Degree Ordering).");
        }
        System.out.println();

        // Backtracking with Warnsdorff's Rule (Knight's Tour - Example on a 5x5 grid)
        System.out.println("4. Backtracking with Warnsdorff's Rule (Knight's Tour - Example):");
        //  For Warnsdorff, the graph needs to represent knight's moves.  We'll use a simple example.
        int startXWarnsdorff = 0;
        int startYWarnsdorff = 0;
        if (hasHamiltonianCycleWarnsdorff(graph, startXWarnsdorff, startYWarnsdorff)) {
            System.out.println("Hamiltonian Cycle exists (Warnsdorff's Rule - Knight's Tour).");
        } else {
            System.out.println("Hamiltonian Cycle does not exist (Warnsdorff's Rule - Knight's Tour).");
        }
        System.out.println();

        // Randomized Backtracking
        System.out.println("5. Randomized Backtracking:");
        if (hasHamiltonianCycleRandomized(graph)) {
            System.out.println("Hamiltonian Cycle exists (Randomized).");
        } else {
            System.out.println("Hamiltonian Cycle does not exist (Randomized).");
        }
        System.out.println();

        scanner.close();
    }
}

